optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/JaxbIO.java

```diff
@@ -52,25 +79,125 @@ public JaxbIO(Class<T> rootClass, int indentation) {
+                throw new OptaPlannerXmlSerializationException(errMessageWithValidationEvents, jaxbException);
+            } else {
+                throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+            }
+        }
+    }
+
+    private Unmarshaller createUnmarshaller() {
+        try {
+            return jaxbContext.createUnmarshaller();
+        } catch (JAXBException e) {
+            String errMessage = String.format("Unable to create JAXB unmarshaller for a root element class (%s).",
+                    rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+    }
+
+    // TODO: provide JavaDoc
+    public T readOverridingNamespace(Reader reader, ElementNamespaceOverride... elementNamespaceOverrides) {
+        Objects.requireNonNull(reader);
+        Objects.requireNonNull(elementNamespaceOverrides);
+
+        final String errMessage = String.format("Unable to read the (%s) from XML with overriding elements' namespaces: %s.",
+                rootClass.getName(), Arrays.toString(elementNamespaceOverrides));
+
+        // Create a SAXParser to use its XMLReader on the XMLFilter
+        SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
+        SAXParser saxParser;
+        try {
+            // Protect the parser against the XXE attack
+            // https://owasp.org/www-project-top-ten/OWASP_Top_Ten_2017/Top_10-2017_A4-XML_External_Entities_(XXE)
+            saxParserFactory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+            saxParser = saxParserFactory.newSAXParser();
+        } catch (ParserConfigurationException | SAXException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+        XMLReader xmlReader;
+        try {
+            xmlReader = saxParser.getXMLReader();
+        } catch (SAXException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+
+        XMLFilter namespaceOverridingXmlFilter = new NamespaceOverridingXmlFilter(xmlReader, elementNamespaceOverrides);
+        namespaceOverridingXmlFilter.setParent(xmlReader);
+
+        // Use UnmarshallerHandler as a content handler for the XML filter.
+        Unmarshaller unmarshaller = createUnmarshaller();
+        UnmarshallerHandler unmarshallerHandler = unmarshaller.getUnmarshallerHandler();
+        namespaceOverridingXmlFilter.setContentHandler(unmarshallerHandler);
+
+        InputSource xmlInputSource = new InputSource(reader);
+        try {
+            // Parse the XML to feed its content into the UnmarshallerHandler.
+            namespaceOverridingXmlFilter.parse(xmlInputSource);
+        } catch (IOException | SAXException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+
+        try {
+            @SuppressWarnings("unchecked")
```

Comment:
Same comment regarding the annotation and the `object` return.

Reply:
See above.